### PR TITLE
Pkcs11 login

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Depends on https://github.com/OP-TEE/optee_os/pull/3750 and https://github.com/OP-TEE/optee_test/pull/418

File: libckteec/include/pkcs11_ta.h [hunk 1]
```diff
@@ -177,6 +177,79 @@ enum pkcs11_ta_cmd {
...
+	 *              32bit session handle,
+	 *              32bit old PIN byte size,
+	 *              32bit new PIN byte size,
+	 *              byte array: PIN data,
+	 *              byte array: new PIN data,
+	 *       ]
+	 * [out] memref[0] = 32bit return code, enum pkcs11_rc
+	 *
+	 * This command relates to the PKCS#11 API function
+	 * C_CloseAllSessions().
```
Comment (etienne-lms) [inline libckteec/include/pkcs11_ta.h:223]: s/C_CloseAllSessions/C_SetPIN/

**Comment by jenswi-linaro** [comment]
Update

File: libckteec/src/pkcs11_api.c [hunk 2]
```diff
@@ -347,16 +359,21 @@ CK_RV C_SetPIN(CK_SESSION_HANDLE hSession,
...
-	if (!lib_initiated())
-		return CKR_CRYPTOKI_NOT_INITIALIZED;
+	if (lib_initiated())
+		rv = ck_set_pin(hSession, pOldPin, ulOldLen, pNewPin, ulNewLen);
 
-	return CKR_FUNCTION_NOT_SUPPORTED;
+	ASSERT_CK_RV(rv, CKR_CRYPTOKI_NOT_INITIALIZED, CKR_DEVICE_ERROR,
+		     CKR_DEVICE_MEMORY, CKR_DEVICE_REMOVED,
+		     CKR_FUNCTION_CANCELED, CKR_FUNCTION_FAILED,
+		     CKR_GENERAL_ERROR, CKR_HOST_MEMORY, CKR_MECHANISM_INVALID,
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_api.c:370]: remove `CKR_MECHANISM_INVALID` from this list.

Comment (jenswi-linaro) [reply libckteec/src/pkcs11_api.c]: OK

File: libckteec/src/pkcs11_api.c [hunk 3]
```diff
@@ -193,15 +193,20 @@ CK_RV C_InitToken(CK_SLOT_ID slotID,
...
+		rv = ck_init_token(slotID, pPin, ulPinLen, pLabel);
 
-	return CKR_FUNCTION_NOT_SUPPORTED;
+	ASSERT_CK_RV(rv, CKR_ARGUMENTS_BAD, CKR_CRYPTOKI_NOT_INITIALIZED,
+		     CKR_DEVICE_ERROR, CKR_DEVICE_MEMORY, CKR_DEVICE_REMOVED,
+		     CKR_FUNCTION_CANCELED, CKR_FUNCTION_FAILED,
+		     CKR_GENERAL_ERROR, CKR_HOST_MEMORY, CKR_OK,
+		     CKR_PIN_INCORRECT, CKR_PIN_LOCKED, CKR_SESSION_EXISTS,
+		     CKR_SLOT_ID_INVALID, CKR_TOKEN_NOT_PRESENT,
+		     CKR_TOKEN_NOT_RECOGNIZED, CKR_TOKEN_WRITE_PROTECTED);
```
Comment (Emantor) [inline libckteec/src/pkcs11_api.c:207]: Is this list based on the spec or based on the return codes of the TA? Per spec `C_InitToken` can also return `CKR_SESSION_HANDLE_INVALID`, `CKR_DEVICE_REMOVED` and `CKR_SESSION_CLOSED`.

Comment (jenswi-linaro) [reply libckteec/src/pkcs11_api.c:207]: This is based on the spec, PKCS #11 Cryptographic Token Interface Base Specification Version 2.40. In fact it's copy paste from the spec and sorted in alphabetical order.
Which spec did you look in?

Comment (Emantor) [reply libckteec/src/pkcs11_api.c:207]: I missed a `expect` in the spec, this fits the spec indeed :facepalm:

Comment (jenswi-linaro) [reply libckteec/src/pkcs11_api.c:207]: Thanks for double checking.

**Comment by jenswi-linaro** [comment]
Squashed, rebased and tag applied.

**Comment by jforissier** [comment]
I'll merge once the checkpatch errors are cleared.

**Comment by Emantor** [comment]
My bad, I forgot the closing `>`…

**Comment by jforissier** [comment]
@Emantor Reply button -> "Create a new saved reply..."

:wink:

**Comment by jenswi-linaro** [comment]
Fixed the tags

**Comment by jenswi-linaro** [comment]
Fixed the other checkpatch error (long line).

**Pull request merged**